src/Dilex.php

```diff
@@ -1,97 +1,329 @@
 
     /**
-     * @param string $pattern
-     * @param Endpoint $to
-     * @return \Silex\Controller
+     * @var AfterWrapper
      */
-    public function match( $pattern, $to = null )
-    {
-        return parent::match( $pattern, $to );
-    }
+    private $afterEventListenerWrapper;
 
     /**
-     * @param string $pattern
-     * @param Endpoint $to
-     * @return \Silex\Controller
+     * @var FinishWrapper
      */
-    public function delete( $pattern, $to = null )
-    {
-        return parent::delete( $pattern, $to );
-    }
+    private $finishEventListenerWrapper;
+
+    /**
+     * @var ErrorWrapper
+     */
+    private $errorEventListenerWrapper;
+
+    /**
+     * @var BeforeControllerListenerRunner
+     */
+    private $beforeControllerListenerRunner;
+
+    /**
+     * @var AfterControllerListenerRunner
+     */
+    private $afterControllerListenerRunner;
+
+    /**
+     * @var StringToResponseListener
+     */
+    private $stringToResponseListener;
 
     /**
-     * @param Middleware|string $middleware
-     * @param int $priority
+     * @var string|null
      */
-    public function before( $middleware, $priority = 0 )
+    private $projectDirectory = null;
+
+    /**
+     * @var string|null
+     */
+    private $cacheDirectory = null;
+
+    /**
+     * @var string|null
+     */
+    private $logDirectory = null;
+
+    public function __construct( string $environment, bool $debug,
+                                 ContainerInterface $fallbackContainerInterface = null )
+    {
+        parent::__construct( $environment, $debug );
+        $this->fallbackContainerInterface = $fallbackContainerInterface;
+        $this->routeRegistry = new RouteRegistry();
+        $this->containerProvider = new ContainerProvider();
+        $this->eventListenerRegistry = new EventListenerRegistry();
+        $this->beforeEventListenerWrapper = new BeforeWrapper( $this->containerProvider );
+        $this->afterEventListenerWrapper = new AfterWrapper( $this->containerProvider );
+        $this->finishEventListenerWrapper = new FinishWrapper( $this->containerProvider );
+        $this->errorEventListenerWrapper = new ErrorWrapper( $this->containerProvider );
+        $this->beforeControllerListenerRunner = new BeforeControllerListenerRunner( $this->containerProvider );
+        $this->afterControllerListenerRunner = new AfterControllerListenerRunner( $this->containerProvider );
+        $this->stringToResponseListener = new StringToResponseListener();
+    }
+
+    public function setProjectDirectory( ?string $projectDirectory ): void
+    {
+        $this->projectDirectory = rtrim( $projectDirectory, '/' );
+    }
+
+    public function setCacheDirectory( ?string $cacheDirectory ): void
+    {
+        $this->cacheDirectory = '/' . trim( $cacheDirectory, '/' ) . '/';
+    }
+
+    public function setLogDirectory( ?string $logDirectory ): void
+    {
+        $this->logDirectory = '/' . trim( $logDirectory, '/' );
```

Comment:
The parameter `$logDirectory` here is nullable, but passing null would result in the directory being set to `"//"`, is that desirable? Same goes for the other two above. I see there are null checks in the getters below, so wondered if the intent was to actually set the properties to null if null was passed in.

Reply:
Thanks @alex1987, indeed these should be null if the parameter passed is null. I commited a fix.